File: backend/moralis.py
import aiohttp
import asyncio
from dotenv import load_dotenv
import os

load_dotenv()
api_key = os.getenv('MORALIS_API_KEY')

async def get_portfolio(address: str):
    url = f'https://solana-gateway.moralis.io/account/mainnet/{address}/portfolio'

    headers = {
        'accept': 'application/json',
        'X-API-Key': api_key,
    }
    
    try:
        async with aiohttp.ClientSession() as session:
            async with session.get(url, headers=headers) as response:
                return await response.json()
            
    except Exception as e:
        print(f"Error fetching wallet portfolio: {e}")
        return None
    
async def get_token_price(address: str):
    url = f'https://solana-gateway.moralis.io/token/mainnet/{address}/price'

    headers = {
        'accept': 'application/json',
        'X-API-Key': api_key,
    }

    try:
        async with aiohttp.ClientSession() as session:
            async with session.get(url, headers=headers) as response:
                res = await response.json()
                return float(res['usdPrice'])
    
    except Exception as e:
        print(f"Error fetching token price: {e}")
        return None
    

async def get_net_worth(address: str):
    portfolio = await get_portfolio(address)
    
    try:
        if not portfolio:
            return None
        
        net_worth = 0
        for token in portfolio['tokens']:
            token_price = await get_token_price(token['mint'])
            
            if token_price:
                net_worth += float(token['amount']) * float(token_price)
        
        return net_worth
    
    except Exception as e:
        print(f"Error calculating net worth: {e}")
        return None

File: backend/moralis.js
require('dotenv').config()

const API_KEY = process.env.MORALIS_API_KEY;

async function getSVMPortfolio(address) {
	try {
		return await (await fetch(`https://solana-gateway.moralis.io/account/mainnet/${address}/portfolio`, {
			headers: {
				"accept": "application/json",
				"X-API-Key": API_KEY
			}
		})).json();
	} catch(err) {
		console.error(err);
		return {
			nativeBalance: {
				solana: "0"
			},
			tokens: []
		};
	}
}

async function getSPLPrice(mint) {
	try {
		return (await (await fetch(`https://solana-gateway.moralis.io/token/mainnet/${mint}/price`, {
			headers: {
				"accept": "application/json",
				"X-API-Key": API_KEY
			}
		})).json()).usdPrice;
	} catch(err) {
		console.error(err);
		return 0;
	}
}

async function getSVMNetWorth(address) {
	try {
		let portfolio = await getSVMPortfolio(address);
		let netWorth = 0;
		let solanaPrice = await getSPLPrice("So11111111111111111111111111111111111111112");
		netWorth += solanaPrice * parseFloat(portfolio.nativeBalance.solana);
		for(let i = 0; i < portfolio.tokens.length; i++) {
			let price = (await getSPLPrice(portfolio.tokens[i].mint)) || 0;
			netWorth += parseFloat(portfolio.tokens[i].amount) * parseFloat(price);
		}
		return netWorth;
	} catch(err) {
		console.error(err);
		return 0;
	}
}

async function getSVMSwaps(address, start) {
	try {
		(await (await fetch(`https://solana-gateway.moralis.io/account/mainnet/${address}/swaps?order=DESC`, {
			headers: {
				"accept": "application/json",
				"X-API-Key": API_KEY
			}
		})).json()).result;
	} catch(err) {
		return [];
	}
}

async function getSVMPNL(address, start) {
	let swaps = await getSVMSwaps(address, start);
	for(let i = 0; i < swaps.length; i++) {

	}
}

(async() => {
	console.log(await getSVMNetWorth(""));
})();